src/main/scala/org/neo4j/spark/LazySessionIterator.scala

```diff
@@ -0,0 +1,32 @@
+package org.neo4j.spark
+
+import org.neo4j.driver.Driver
+import org.neo4j.driver.Session
+
+class LazySessionIterator[T](iterator: Iterator[T], driver: Driver, session: Session) extends Iterator[T] {
+  private lazy val it = iterator
+
+  override def hasNext: Boolean = {
+    val hasNext = it.hasNext
+    if (!hasNext) {
+      close
+    }
+    hasNext
+  }
+
+  private def close = {
+    if (session.isOpen) session.close()
```

Comment:
null check on session and wrap with try-catch in case session.close throws  an exception